## Capstone Project
## Project:Medical Representative Smart Helper 

Medical representatives are the key point of contact between
pharmaceutical and medical companies and healthcare professionals.
A medical representative's job is to promote and sell their company's
products, whether that's pharmaceutical drugs or medical equipment.
Customers can include doctors, nurses and pharmacists. The medical
representative will increase product awareness, answer queries,
company.

The job of medical representative has a lot of challenges but the most
important one is how the medical representative convinces a doctor to write a
drug from the medical company which he works for in his prescription where
there are many competitors' drugs that have the same active ingredient from
other companies.
The goal is to build a model that will predict if a doctor with some
specific features will write or not one of 6 drugs for a medical
company that has license to produce several types of local produced
drugs with the same active ingredients of multinational original drugs.

## Getting Started

In this project   The solution will be prediction if a doctor will write one of the 6
drugs in his prescription or not to help medical representative to
choose the right doctor to convince him to write the drug that he
represent according to the features of the doctor. First I will use
'pandas' to split the training data to features and targets and use
'matplotlib' to do some visualization of the data to get some
understanding.
Second I will split my data to 80% for training and 20% for validation
testing.
For training model I will use supervised machine learning to build my
model using sklearn. I will use 'decision tree classifiers' and
'adaboost classifier' to build my model. Finally I will select the
best model with best accuracy.
At last I will use final testing data that will be separated from the
training and validation data to make my final test and get the accuracy.

## Exploring the Data
In this part i'm going to explore the data and all the features.

### Datasets and Inputs
The dataset was collected by me in a year from private clinics,
hospitals and pharmacies
All data are public data that was collected from doctors and
pharmacists about 6 drugs produced by a certain medical company.


In [21]:
import numpy as np
import pandas as pd
from IPython.display import display # Allows the use of display() for DataFrames


# Pretty display for notebooks
%matplotlib inline

# Load the dataset
full_data = pd.read_csv("medical.txt")

# Print the first few entries of the medical representative
display(full_data.head())
print(full_data.shape)
'''
print(max(full_data['price']))
print(max(full_data['exam_price']))
'''

medicine   area speciality  price dr_class  exam_price clinic_hos  write
0    type1  area1      chest     45        a         200     clinic      1
1    type4  area2         im     36        b         100     clinic      1
2    type1  area8      chest     45        a          75   hospital      1
3    type1  area5      chest     45        a          30   hospital      1
4    type5  area6        uro     29        a         220     clinic      0

(390, 8)


"\nprint(max(full_data['price']))\nprint(max(full_data['exam_price']))\n"

## Implementation: Data Exploration
A cursory investigation of the dataset will determine how many individuals fit into either group, and will tell us about the percentage of doctors who wrote any of the 6 drugs in their prescription. In the code cell below, i will compute the following:

The total number of records, 'n_records'

The number of doctors who wrote any of the 6 drugs in their prescription, 'n_doctor_who_write'.

The number of doctors who didn't write any of the 6 drugs in their prescription, 'n_doctor_who_not_write'.

The percentage of doctors who wrote any of the 6 drugs in their prescription , 'write_percent'.

In [2]:
#  Total number of records
n_records = full_data.shape[0]
# The number of doctors who wrote any of the 6 drugs in their prescription
n_doctor_who_write =full_data[full_data.write==1].shape[0]


# The number of doctors who didn't write any of the 6 drugs in their prescription
n_doctor_who_not_write =full_data[full_data.write==0].shape[0]


# The percentage of doctors who wrote any of the 6 drugs in their prescription
write_percent = (n_doctor_who_write/n_records)*100
#Minimum price of drugs
minimum_drug_price = min(full_data['price'])

#Maximum price of the data
maximum_drug_price = max(full_data['price'])

#Mean price of the data
mean_drug_price = np.array(full_data['price']).mean()
#Median price of the data
median_drug_price  =np.median(full_data['price'])

#Standard deviation of prices of the data
std_drug_price =np.array(full_data['price']).std()


#Minimum price of examination price
minimum_exam_price = min(full_data['exam_price'])

#Maximum price of the data
maximum_exam_price = max(full_data['exam_price'])

#Mean price of the data
mean_exam_price = np.array(full_data['exam_price']).mean()
#Median price of the data
median_exam_price  =np.median(full_data['exam_price'])

#Standard deviation of prices of the data
std_exam_price =np.array(full_data['exam_price']).std()
# Show the calculated statistics
print("Statistics for drugs prices :\n")
print("Minimum drug price: ${}".format(minimum_drug_price)) 
print("Maximum drug price: ${}".format(maximum_drug_price))
print("Mean drug prices: ${}".format(mean_drug_price))
print("Median drug prices ${}".format(median_drug_price))
print("Standard deviation of prices: ${}\n".format(std_drug_price))

print("Statistics for examination prices :\n")
print("Minimum examination price: ${}".format(minimum_exam_price)) 
print("Maximum examination price: ${}".format(maximum_exam_price))
print("Mean examination price: ${}".format(mean_exam_price))
print("Median examination price ${}".format(median_exam_price))
print("Standard deviation of examination prices: ${}\n".format(std_exam_price))
print("Statistics for doctors writing :\n")
print("Total number of records: {}".format(n_records))
print("Doctors who write the copmany drugs: {}".format(n_doctor_who_write))
print("Doctors who didn't write the company drugs: {}".format(n_doctor_who_not_write))
print("Percentage of doctor who write: {}%".format(write_percent))

Statistics for drugs prices :

Minimum drug price: $20
Maximum drug price: $45
Mean drug prices: $35.715384615384615
Median drug prices $36.0
Standard deviation of prices: $8.740035746612882

Statistics for examination prices :

Minimum examination price: $30
Maximum examination price: $350
Mean examination price: $121.2051282051282
Median examination price $80.0
Standard deviation of examination prices: $86.61858019639882

Statistics for doctors writing :

Total number of records: 390
Doctors who write the copmany drugs: 229
Doctors who didn't write the company drugs: 161
Percentage of doctor who write: 58.71794871794872%


## Featureset Exploration
### Input Data fields (Features and Labels)
The input data consists of 390 samples collected from 253 doctors.
This data will be used for training and validation test in a file
named medical attached with the project proposal.
Every sample consists of 7 features and 1 label(the output target).
* **medicine:** The commercial name of the medicine that I refer to it as type1 to type6.
* **Price:** price for every drug of these 6 drugs for patient.
* **exam_price:** examination price that the doctor takes from the patient.
* **clinic_hos:** classification if this doctor has a private 'clinic' or works in a 'hospital'.
* **dr_class:** doctor class was classified according to number of patients for every doctor and how much is this doctor famous and it was classified to 'a' and 'b'.

**where:**

* **class a:** are doctors with high patient rate
* **class b:** are doctors with medium patient rate
* **area:** every medical representative cover some areas in a city where the medical representative has to make a lot of visits for alot of doctors in different places and I refer to each area with the name (area1 to area8)with total 8 areas that I have covered to make my prospection.
* **specialty:** the specialty of the doctor is one of the most important features because it will determine which of the 6 drugs will be suitable for this doctor to write it in his prescription. 

**These specialties are:**
* **Chest:** chest specialist
* **Im:** internal medicine specialist
* **Cd:** cardiology specialist
* **Neuro:** neurology specialist
* **Git:** gastrointestinal tract specialist
* **Ent:** ear, nose and throat specialist
* **Sur:** surgery specialist
* **Uro:** urology specialist
* **Gp:** general practitioner
* **Or:** orthopedic specialist
* **Vas:** vascular specialst

**With total 11 specialties.**

* **Target (label):** just one target to classify if this doctor write any of these drugs in prescription to his patients or not. With (1) if the doctor writes the drug in his prescription and (0)if not.

### Metrics and the Naive Predictor

I'm going to use both of Accuracy and F1-score to evaluate the model. 

**Accuracy:** is a common metric for binary classifiers it takes into account both true positives and true negatives 
**where:**

**True positive (TP):** is the number of correct predictions that an instance is true. In our dataset the TP is the number of doctors who actually wrote one of our 6 drugs in their patient prescription.

**True negative(TN):** is presenting a number of correct predictions that an instance is false. In our dataset the TP is the number of doctors who actually didn't wrote any one of our 6 drugs in their patient prescription.

accuracy = (true positives + true negatives)/all data

$$ F_{\beta} = (1 + \beta^2) \cdot \frac{precision \cdot recall}{\left( \beta^2 \cdot precision \right) + recall} $$

In particular, when $\beta = 0.5$, more emphasis is placed on precision. This is called the **F$_{0.5}$ score** (or F-score for simplicity).


####  accuracy, precision, recall

** Accuracy ** measures how often the classifier makes the correct prediction. It’s the ratio of the number of correct predictions to the total number of predictions (the number of test data points).

** Precision ** tells us what proportion of messages we classified as spam, actually were spam.
It is a ratio of true positives(words classified as spam, and which are actually spam) to all positives(all words classified as spam, irrespective of whether that was the correct classificatio), in other words it is the ratio of

`[True Positives/(True Positives + False Positives)]`

** Recall(sensitivity)** tells us what proportion of messages that actually were spam were classified by us as spam.
It is a ratio of true positives(words classified as spam, and which are actually spam) to all the words that were actually spam, in other words it is the ratio of

`[True Positives/(True Positives + False Negatives)]`



## Benchmark Model:
Because the dataset was collected by myself ,there is no past Benchmark model to compare its performance with my model.
I'm going to use a naive prediction  to help establish a benchmark for whether a model is performing well.


## Naive Predictor Performace:
* the purpose of generating a naive predictor is simply to show what a base model without any intelligence would look like. In the real world, ideally your base model would be either the results of a previous model or could be based on a research paper upon which you are looking to improve. When there is no benchmark model set, getting a result better than random choice is a place you could start from.


* because i want to build a model that predict who are the doctors who will write any of the six drugs , i'm going to assume that all doctors are goint to write and calcuate both of accuracy and F1_score to get the model prerfomance in the next code

In [3]:
TP=np.sum(full_data['write'])
FP= n_records - TP
TN=0
FN=0
accuracy = TP/(TP+FP)
recall = TP/(TP+FN)
precision = TP/(TP+FP)
 
# i will Calculate F-score using the formula above for beta = 0.5 and correct values for precision and recall.
beta = 0.5
fscore =(1+ beta**2)*(precision * recall)/((beta**2*precision)+recall) 


# Print the results 
print("Naive Predictor: [Accuracy score: {:.4f}, F-score: {:.4f}]".format(accuracy, fscore))

Naive Predictor: [Accuracy score: 0.5872, F-score: 0.6400]


* **The accuracy score is :** 0.5872
* **The F-score is :**0.6400

* these two valuse are my start point where i have to build a model with decision tree 'classifiers' or 'adaboost classifier' that has better valuse for accuracy and F_score than thses values

## Preparing the Data
Before data can be used as input for machine learning algorithms, it often must be cleaned, formatted, and restructured — this is typically known as **preprocessing**. Fortunately, for this dataset, there are no invalid or missing entries we must deal with, however, there are some qualities about certain features that must be adjusted. This preprocessing can help tremendously with the outcome and predictive power of nearly all learning algorithms.


### Implementation: Data Preprocessing

From the table in Exploring the Data above, we can see there are several features for each record that are non-numeric. Typically, learning algorithms expect input to be numeric, which requires that non-numeric features (called categorical variables) be converted. One popular way to convert categorical variables is by using the one-hot encoding scheme. One-hot encoding creates a "dummy" variable for each possible category of each non-numeric feature. For example, assume someFeature has three possible entries: A, B, or C. We then encode this feature into someFeature_A, someFeature_B and someFeature_C.

**where:**
* **someFeature_A will be= 1 0 0**
* **someFeature_B will be= 0 1 0**
* **someFeature_c will be= 0 0 1**  

* **I'm going to use pandas.get_dummies() to perform one-hot encoding on the 'full_data' data.**

* **I don't need to Convert the target label 'write' to numerical because it is numerical where the value (1) represent 'write' and the value (0) represent 'not write'**


In [4]:
one_hot_data=pd.get_dummies(full_data)
one_hot_data

price  exam_price  write  medicine_type1  medicine_type2  medicine_type3  \
0       45         200      1               1               0               0   
1       36         100      1               0               0               0   
2       45          75      1               1               0               0   
3       45          30      1               1               0               0   
4       29         220      0               0               0               0   
..     ...         ...    ...             ...             ...             ...   
385     29          75      0               0               0               0   
386     30          70      0               0               1               0   
387     45         120      1               1               0               0   
388     45          75      1               1               0               0   
389     45         220      1               1               0               0   

     medicine_type4  medicine_type5  medicine_type6  area_area1  ...  \
0                 0               0               0           1  ...   
1                 1               0               0           0  ...   
2                 0               0               0           0  ...   
3                 0               0               0           0  ...   
4                 0               1               0           0  ...   
..              ...             ...             ...         ...  ...   
385               0               1               0           0  ...   
386               0               0               0           0  ...   
387               0               0               0           0  ...   
388               0               0               0           0  ...   
389               0               0               0           1  ...   

     speciality_im  speciality_neuro  speciality_or  speciality_sur  \
0                0                 0              0               0   
1                1                 0              0               0   
2                0                 0              0               0   
3                0                 0              0               0   
4                0                 0              0               0   
..             ...               ...            ...             ...   
385              0                 0              0               1   
386              1                 0              0               0   
387              1                 0              0               0   
388              0                 0              0               1   
389              0                 0              0               0   

     speciality_uro  speciality_vas  dr_class_a  dr_class_b  \
0                 0               0           1           0   
1                 0               0           0           1   
2                 0               0           1           0   
3                 0               0           1           0   
4                 1               0           1           0   
..              ...             ...         ...         ...   
385               0               0           0           1   
386               0               0           0           1   
387               0               0           1           0   
388               0               0           0           1   
389               0               0           1           0   

     clinic_hos_clinic  clinic_hos_hospital  
0                    1                    0  
1                    1                    0  
2                    0                    1  
3                    0                    1  
4                    1                    0  
..                 ...                  ...  
385                  0                    1  
386                  0                    1  
387                  1                    0  
388                  0                    1  
389                  1                    0  

[390

**Now all the data had been turned to numerical values but i need all the numerical data to be in range from 0 to 1, so i'm going to scale both of 'price' and 'exam_price' features to be in range from (0 to 1) by dividing both of them by their maximum values**

In [5]:

# Import sklearn.preprocessing.StandardScaler
from sklearn.preprocessing import MinMaxScaler

# Initialize a scaler, then apply it to the features
scaler = MinMaxScaler() # default=(0, 1)
numerical = ['price','exam_price']

one_hot_data[numerical] = scaler.fit_transform(one_hot_data[numerical])

# Show an example of a record with scaling applied
display(one_hot_data.head(n = 5))

price  exam_price  write  medicine_type1  medicine_type2  medicine_type3  \
0   1.00    0.531250      1               1               0               0   
1   0.64    0.218750      1               0               0               0   
2   1.00    0.140625      1               1               0               0   
3   1.00    0.000000      1               1               0               0   
4   0.36    0.593750      0               0               0               0   

   medicine_type4  medicine_type5  medicine_type6  area_area1  ...  \
0               0               0               0           1  ...   
1               1               0               0           0  ...   
2               0               0               0           0  ...   
3               0               0               0           0  ...   
4               0               1               0           0  ...   

   speciality_im  speciality_neuro  speciality_or  speciality_sur  \
0              0                 0              0               0   
1              1                 0              0               0   
2              0                 0              0               0   
3              0                 0              0               0   
4              0                 0              0               0   

   speciality_uro  speciality_vas  dr_class_a  dr_class_b  clinic_hos_clinic  \
0               0               0           1           0                  1   
1               0               0           0           1                  1   
2               0               0           1           0                  0   
3               0               0           1           0                  0   
4               1               0           1           0                  1   

   clinic_hos_hospital  
0                    0  
1                    0  
2                    1  
3                    1  
4                    0  

[5 rows x 32 columns]

In [6]:
outcomes = one_hot_data['write']
features= one_hot_data.drop('write', axis = 1)
# Show the new dataset with 'write' removed
display(features.head())

price  exam_price  medicine_type1  medicine_type2  medicine_type3  \
0   1.00    0.531250               1               0               0   
1   0.64    0.218750               0               0               0   
2   1.00    0.140625               1               0               0   
3   1.00    0.000000               1               0               0   
4   0.36    0.593750               0               0               0   

   medicine_type4  medicine_type5  medicine_type6  area_area1  area_area2  \
0               0               0               0           1           0   
1               1               0               0           0           1   
2               0               0               0           0           0   
3               0               0               0           0           0   
4               0               1               0           0           0   

   ...  speciality_im  speciality_neuro  speciality_or  speciality_sur  \
0  ...              0                 0              0               0   
1  ...              1                 0              0               0   
2  ...              0                 0              0               0   
3  ...              0                 0              0               0   
4  ...              0                 0              0               0   

   speciality_uro  speciality_vas  dr_class_a  dr_class_b  clinic_hos_clinic  \
0               0               0           1           0                  1   
1               0               0           0           1                  1   
2               0               0           1           0                  0   
3               0               0           1           0                  0   
4               1               0           1           0                  1   

   clinic_hos_hospital  
0                    0  
1                    0  
2                    1  
3                    1  
4                    0  

[5 rows x 31 columns]

**data preprocessing has been completed and now i will split the data in to training data and validation data with 80% of the data as training and 20% for validation data**

In [7]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(features, outcomes, test_size=0.2, random_state=42)

### Features vs Targets

In [8]:
features = np.array(X_train)
targets = np.array( y_train)
features_valid = np.array( X_test)
targets_valid = np.array(y_test)

**i has split both of training data and validation data into features and targets**

**where:**

features :       training features

targets:         training targets

features_valid : validation features

targets_valid:   validation targets  


### Implementation: Initial Model Evaluation

In [9]:
from sklearn.tree import DecisionTreeClassifier
#define a Decision Tree Classifier
model = DecisionTreeClassifier(random_state=42)
model=model.fit(features, targets)
model

DecisionTreeClassifier(random_state=42)

In [10]:
#y_train_pred = model.predict(features)
from sklearn.metrics import fbeta_score, accuracy_score
y_valid_pred = model.predict(features_valid)

# Calculate the accuracy
from sklearn.metrics import accuracy_score
#train_accuracy = accuracy_score(targets, y_train_pred)
valid_accuracy = accuracy_score(targets_valid, y_valid_pred)

#print('The training accuracy is', train_accuracy)
print("Unoptimized model\n------")
print('The validation accuracy is', valid_accuracy)
print("F-score on validation data: {:.4f}".format(fbeta_score(targets_valid, y_valid_pred, beta = 0.5)))


Unoptimized model
------
The validation accuracy is 0.7435897435897436
F-score on validation data: 0.7819


In [11]:
from sklearn.ensemble import AdaBoostClassifier


clf = AdaBoostClassifier(random_state=42)
clf=clf.fit(features, targets)
predictions = clf.predict(features_valid)
print("Unoptimized model\n------")
print("Accuracy score on validaton data: {:.4f}".format(accuracy_score(targets_valid, predictions)))
print("F-score on validation data: {:.4f}".format(fbeta_score(targets_valid, predictions, beta = 0.5)))
clf

Unoptimized model
------
Accuracy score on validaton data: 0.6795
F-score on validation data: 0.7255


AdaBoostClassifier(random_state=42)

* After training both of them without optimizing the hyper parameters , i  selected the best of them which gave me the best validation accuracy and the best F_score and that was the 'adaboost' model **where:**

**The adaboost model gave me:**  
* Accuracy score on validation data: 0.7949
* F-score on validation data: 0.8015

**The 'decision tree model gave me :**
* Accuracy score on validation data: 0 .73
* F-score on validation data: 0.79


### Improving Results
After choosing the best Un optimized model that was 'AdaBoostClassifier' , I  tuned it's  hyper parameters to improve the accuracy and F_score results
•	I used sklearn.grid_search to import GridSearchCV to find the optimal parameters from these parameters: 

•	'n_estimators':  [75,100,200,300]

•	 'learning_rate': [0.1,0.2,0.4,0.5,0.7,1.0]

•	To get the best optimized model I used sklearn.metrics to  import make_scorer to get the best F_score to the best optimized model that will be my final model


### Implementation: Model Tuning
I will Use grid search (`GridSearchCV`) with at least one important parameter tuned with at least 3 different values. I will need to use the entire training set for this. In the code cell below, i will need to implement the following:
- Import [`sklearn.grid_search.GridSearchCV`](http://scikit-learn.org/0.17/modules/generated/sklearn.grid_search.GridSearchCV.html) and [`sklearn.metrics.make_scorer`](http://scikit-learn.org/stable/modules/generated/sklearn.metrics.make_scorer.html).
- Initialize the classifier you've chosen and store it in `clf`.
- Create a dictionary of parameters you wish to tune for the chosen model.
 - Example: `parameters = {'parameter' : [list of values]}`.
 - **Note:** Avoid tuning the `max_features` parameter of your learner if that parameter is available!
- Use `make_scorer` to create an `fbeta_score` scoring object (with $\beta = 0.5$).
- Perform grid search on the classifier `clf` using the `'scorer'`, and store it in `grid_obj`.
- Fit the grid search object to the training data (`X_train`, `y_train`), and store it in `grid_fit`.

In [12]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import fbeta_score,accuracy_score
from sklearn.model_selection import ShuffleSplit
# Initialize the classifier
print("step1")
clf = DecisionTreeClassifier(random_state=42)
print("step2")
# Create the parameters list you wish to tune, using a dictionary if needed.
parameters = {'max_depth':[11,12],'min_samples_leaf':[1,5], 'min_samples_split':[2,3]}
#parameters = {'max_depth':[1,2,3,4,5,6,7,8,9,10,11,12,15,20],'min_samples_leaf':[2,4,6,8,10,1,15,20,30,40], 'min_samples_split':[2,3,4,5,6,7,8,9,10,15,20,30,40]}
#parameters = {'n_estimators': [75,100,200,300], 'learning_rate': [0.1,0.2,0.4,0.5,0.7,1.0]}
print("step3")
# Make an fbeta_score scoring object using make_scorer()
scorer = make_scorer(fbeta_score, beta = 0.5)
cv_sets=ShuffleSplit(n_splits=10, random_state=42, test_size=0.2, train_size=None)
print("step4")
# Perform grid search on the classifier using 'scorer' as the scoring method using GridSearchCV()
grid_obj = GridSearchCV(clf, parameters,scoring=scorer,cv=cv_sets)
print("step5")
# Fit the grid search object to the training data and find the optimal parameters using fit()
grid_fit =grid_obj.fit(features, targets) 
print("step6")
# Get the estimator
best_clf = grid_fit.best_estimator_
print("step7")
# Make predictions using the unoptimized and model
predictions = (clf.fit(features, targets)).predict(features_valid)
print("step8")
best_predictions = best_clf.predict(features_valid)

# Report the before-and-afterscores
print("Unoptimized model\n------")
print("Accuracy score on validaton data: {:.4f}".format(accuracy_score(targets_valid, predictions)))
print("F-score on validation data: {:.4f}".format(fbeta_score(targets_valid, predictions, beta = 0.5)))
print("\nOptimized Model\n------")
print("Final accuracy score on the validation data: {:.4f}".format(accuracy_score(targets_valid, best_predictions)))
print("Final F-score on the validation data: {:.4f}".format(fbeta_score(targets_valid, best_predictions, beta = 0.5)))
best_clf

step1
step2
step3
step4
step5
step6
step7
step8
Unoptimized model
------
Accuracy score on validaton data: 0.7436
F-score on validation data: 0.7819

Optimized Model
------
Final accuracy score on the validation data: 0.7692
Final F-score on the validation data: 0.8025


DecisionTreeClassifier(max_depth=12, random_state=42)

In [13]:
print("max depth:",best_clf.max_depth)
print("min samples split:",best_clf.min_samples_split)
print("min samples leaf:",best_clf.min_samples_leaf)

max depth: 12
min samples split: 2
min samples leaf: 1


In [14]:
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import make_scorer
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import fbeta_score,accuracy_score
from sklearn.model_selection import ShuffleSplit
# Initialize the classifier
print("step1")
clf = AdaBoostClassifier(random_state=42)
print("step2")
# Create the parameters list you wish to tune, using a dictionary if needed.
parameters = {'n_estimators': [1,2,3,5,10], 'learning_rate': [0.00001,0.0001,0.001,0.01,0.05,0.1,0.2,0.3,0.4,0.5,0.7],'base_estimator': [DecisionTreeClassifier(max_depth=12)]}
#parameters = {'n_estimators': [75,100,200,300], 'learning_rate': [0.1,0.2,0.4,0.5,0.7,1.0]}
print("step3")
# Make an fbeta_score scoring object using make_scorer()
scorer = make_scorer(fbeta_score, beta = 0.5)
#cv_sets=ShuffleSplit(n_splits=10, random_state=42, test_size=0.2, train_size=None)
print("step4")
# Perform grid search on the classifier using 'scorer' as the scoring method using GridSearchCV()
grid_obj = GridSearchCV(clf, parameters,scoring=scorer)
print("step5")
# Fit the grid search object to the training data and find the optimal parameters using fit()
grid_fit =grid_obj.fit(features, targets) 
print("step6")
# Get the estimator
best_clf = grid_fit.best_estimator_
print("step7")
# Make predictions using the unoptimized and model
predictions = (clf.fit(features, targets)).predict(features_valid)
print("step8")
best_predictions = best_clf.predict(features_valid)

# Report the before-and-afterscores
print("Unoptimized model\n------")
print("Accuracy score on validaton data: {:.4f}".format(accuracy_score(targets_valid, predictions)))
print("F-score on validation data: {:.4f}".format(fbeta_score(targets_valid, predictions, beta = 0.5)))
print("\nOptimized Model\n------")
print("Final accuracy score on the validation data: {:.4f}".format(accuracy_score(targets_valid, best_predictions)))
print("Final F-score on the validation data: {:.4f}".format(fbeta_score(targets_valid, best_predictions, beta = 0.5)))
best_clf

step1
step2
step3
step4
step5
step6
step7
step8
Unoptimized model
------
Accuracy score on validaton data: 0.6795
F-score on validation data: 0.7255

Optimized Model
------
Final accuracy score on the validation data: 0.7692
Final F-score on the validation data: 0.8025


AdaBoostClassifier(base_estimator=DecisionTreeClassifier(max_depth=12),
                   learning_rate=1e-05, n_estimators=2, random_state=42)

### Model Evaluation and Validation
* The optimized model's validation_accuracy = 0.7436
* The optimized model's F-score for validation data= 0.7627

For sure it is better than the Un optimized model accuracy = 0.7051  and Un optimized model F-score= 0.7366



### Preprocessing for testing data
i'm going to make data preprocessing for the testing data exactly as i did for the origianl data

In [15]:
testing_data = pd.read_csv("testing.txt")

# Print the first few entries of the RMS Titanic data
display(testing_data.head())
print(testing_data.shape)
print(max(testing_data['price']))
print(max(testing_data['exam_price']))

medicine   area speciality  price dr_class  exam_price clinic_hos  write
0    type1  area1      chest     45        a         220     clinic      1
1    type1  area4         im     45        b          90   hospital      1
2    type1  area2         gp     45        b          60   hospital      1
3    type1  area3         im     45        a         180     clinic      0
4    type4  area8      chest     36        a          75   hospital      0

(30, 8)
45
350


In [16]:
one_hot_testing_data=pd.get_dummies(testing_data)
one_hot_testing_data

price  exam_price  write  medicine_type1  medicine_type2  medicine_type3  \
0      45         220      1               1               0               0   
1      45          90      1               1               0               0   
2      45          60      1               1               0               0   
3      45         180      0               1               0               0   
4      36          75      0               0               0               0   
5      30         150      1               0               1               0   
6      30         200      1               0               1               0   
7      30         165      1               0               1               0   
8      30          90      1               0               1               0   
9      45          30      1               1               0               0   
10     24          90      1               0               0               1   
11     24         340      0               0               0               1   
12     24          50      1               0               0               1   
13     24         100      0               0               0               1   
14     45         120      1               1               0               0   
15     36         350      0               0               0               0   
16     36          50      1               0               0               0   
17     36         180      0               0               0               0   
18     36         200      0               0               0               0   
19     45         100      0               1               0               0   
20     29         100      0               0               0               0   
21     29          50      1               0               0               0   
22     29         250      0               0               0               0   
23     29         170      1               0               0               0   
24     20         130      1               0               0               0   
25     20         290      0               0               0               0   
26     20         150      1               0               0               0   
27     20          50      0               0               0               0   
28     30         100      1               0               1               0   
29     45         160      1               1               0               0   

    medicine_type4  medicine_type5  medicine_type6  area_area1  ...  \
0                0               0               0           1  ...   
1                0               0               0           0  ...   
2                0               0               0           0  ...   
3                0               0               0           0  ...   
4                1               0               0           0  ...   
5                0               0               0           0  ...   
6                0               0               0           1  ...   
7                0               0               0           0  ...   
8                0               0               0           0  ...   
9                0               0               0           0  ...   
10               0               0               0           0  ...   
11               0               0               0           1  ...   
12               0               0               0           0  ...   
13               0               0               0           0  ...   
14               0               0               0           0  ...   
15               1               0               0           1  ...   
16               1               0               0           0  ...   
17               1               0               0           0  ...   
18               1               0               0           0  ...   
19               0               0               0           0  ...   
20               0              

In [17]:


# Initialize a scaler, then apply it to the features
numerical = ['price','exam_price']

one_hot_testing_data[numerical] = scaler.transform(one_hot_testing_data[numerical])

# Show an example of a record with scaling applied
display(one_hot_data.head(n = 5))

price  exam_price  write  medicine_type1  medicine_type2  medicine_type3  \
0   1.00    0.531250      1               1               0               0   
1   0.64    0.218750      1               0               0               0   
2   1.00    0.140625      1               1               0               0   
3   1.00    0.000000      1               1               0               0   
4   0.36    0.593750      0               0               0               0   

   medicine_type4  medicine_type5  medicine_type6  area_area1  ...  \
0               0               0               0           1  ...   
1               1               0               0           0  ...   
2               0               0               0           0  ...   
3               0               0               0           0  ...   
4               0               1               0           0  ...   

   speciality_im  speciality_neuro  speciality_or  speciality_sur  \
0              0                 0              0               0   
1              1                 0              0               0   
2              0                 0              0               0   
3              0                 0              0               0   
4              0                 0              0               0   

   speciality_uro  speciality_vas  dr_class_a  dr_class_b  clinic_hos_clinic  \
0               0               0           1           0                  1   
1               0               0           0           1                  1   
2               0               0           1           0                  0   
3               0               0           1           0                  0   
4               1               0           1           0                  1   

   clinic_hos_hospital  
0                    0  
1                    0  
2                    1  
3                    1  
4                    0  

[5 rows x 32 columns]

In [18]:

final_features_test = np.array(one_hot_testing_data.drop('write', axis=1))
final_targets_test = np.array(one_hot_testing_data['write'])

print(final_features_test.shape)

(30, 31)


**Now i have the testing preprocessed data split into final_features_test and final_targets_test**
* **I will test my final model performance with the testing data**

In [19]:
predictions = (clf.fit(features, targets)).predict(final_features_test)

print("Final accuracy score on the testing data: {:.4f}".format(accuracy_score(final_targets_test, predictions)))
print("Final F-score on the testing data: {:.4f}".format(fbeta_score(final_targets_test, predictions, beta = 0.5)))

Final accuracy score on the testing data: 0.7333
Final F-score on the testing data: 0.7456


In [20]:


best_predictions = best_clf.predict(final_features_test)

print("Final accuracy score on the testing data: {:.4f}".format(accuracy_score(final_targets_test, best_predictions)))
print("Final F-score on the testing data: {:.4f}".format(fbeta_score(final_targets_test, best_predictions, beta = 0.5)))

Final accuracy score on the testing data: 0.7667
Final F-score on the testing data: 0.7843


the result still very good with 
* Final accuracy score on the testing data: 0.7333
* Final F-score on the testing data: 0.7456
Now I'm very sure that my model has done its job
